x-pack/plugins/proxy/server/cluster_doc.ts

```diff
@@ -0,0 +1,201 @@
+  public getNodeForResource(resource: string) {
+    return this.routingTable[resource];
+  }
+
+  public async assignResource(resource: string, data: RoutingNode) {
+    this.routingTable[resource] = data;
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  public async unassignResource(resource: string) {
+    delete this.routingTable[resource];
+    const nodes = await this.getNodeList();
+    const currentTime = new Date().getTime();
+    await this.updateNodeList(this.updateLocalNode(nodes, currentTime));
+  }
+
+  private setConfig(config: ProxyPluginType) {
+    this.updateInterval = config.updateInterval;
+    this.timeoutThreshold = config.timeoutThreshold;
+  }
+
+  private setTimer() {
+    if (this.timer) return;
+    this.log.debug('Set timer to updateNodeMap');
+    this.timer = setInterval(async () => {
+      this.log.debug('Updating node map');
+      await this.mainLoop();
+    }, this.updateInterval);
+  }
+
+  private updateRoutingTable(routingTable: { [key: string]: RoutingNode }): void {
+    const currentRoutes = [...Object.keys(this.routingTable)];
+    for (const [key, node] of Object.entries(routingTable)) {
+      this.routingTable[key] = node;
+      const idx = currentRoutes.findIndex(k => k === key);
+      if (idx) currentRoutes.splice(idx, 1);
+    }
+
+    for (const key of currentRoutes.values()) {
+      delete this.routingTable[key];
+    }
+  }
+
+  private async getNodeList(): Promise<NodeList> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      _source: true,
+    };
+    const data: ClusterDoc = await client.callAsInternalUser('get', params);
+    this.updateRoutingTable(data.routing_table);
+    const nodes: NodeList = data.nodes;
+    return nodes;
+  }
+
+  private async updateNodeList(nodes: NodeList): Promise<void> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const doc = {
+      nodes,
+      routing_table: this.routingTable,
+    };
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      doc,
+    };
+    await client.callAsInternalUser('update', params);
+  }
+
+  private updateLocalNode(nodes: NodeList, finishTime: number): NodeList {
+    nodes[this.nodeName] = {
+      lastChangedTime: finishTime,
+      lastUpdate: finishTime,
+    };
+    return nodes;
+  }
+
+  private async mainLoop(): Promise<any> {
+    const nodes = await this.getNodeList();
+    const finishTime = new Date().getTime();
+
+    for (const [key, node] of Object.entries(nodes)) {
+      if (!node || finishTime - node.lastUpdate > this.timeoutThreshold) {
```

Comment:
I think this part is still different from the design document. `lastChangedTime` has never been used actually.
Also, `lastChangedTime` should be local, don't need to save into elasticsearch.

Reply:
I don't think `lastChangedTime` is necessary (I have removed it).

Since each node is unable to verify that another node actually exists, we only need to look at the `lastUpdate` time, which is the last time this specific node updated the cluster doc. If the node hasn't updated the doc in whatever the `timeoutThreshold` is, we can drop it from the index. I could very well be mis-understanding the point of `lastChangedTime` though

Reply:
The purpose of "lastChangedDate" is to eliminate the impact of inconsistent clocks on individual kibana nodes.  Consider a scenario where in a kibana cluster, the clock of a node is a few minutes faster than other nodes. This time difference may be greater than the "timeoutThreshold". If only based on the "lastUpdate" reported by each node,  When this node runs "mainloop", he incorrectly believes that other nodes are offline due to timeouts.

Reply:
@toddself do you think we should have another meeting?

Reply:
@zfy0701 i think it makes sense to find some time to discuss this.

i don't think we should be in the business of accounting for a wide system clock skew between kibana instances -- there should be an expectation that the system clocks are close enough to account for at most a few seconds of drift at most, which would still be within the default update threshold of 15 seconds.

however, the 15 second update threshold plus [fuyao's comment](https://github.com/elastic/kibana/pull/36907#pullrequestreview-247908766) on elastic's concurrency control has me a bit worried.  the faster we update the nodes, the more likely we are to run into conflicts trying to update the shared document in elasticsearch; anything larger than a handful of nodes is going to start conflicting hard to the point where we'll be spending a large portion of time spinning on retries causing deadlocks in the system.

prior to joining elastic i worked on two systems that had to update an es server with frequent updates and we ran into issues in both places with frequent consistent updates of services.

Reply:
Forgot to discuss this issue in last zoom meeting
I think the correctness of liveness is very important, otherwise e.g. if one node has higher clock could make other node in the cluster disappear. All mission critical system should handle the clock correctly like es, all other database.

We could file an issue and do it in a separate PR that's more feasible.

Reply:
the release notes should mention that clock skew is a potential issue here and that system administrators should endeavor to make sure their clocks are synchronized correctly. the only option we have is to widen the gap between check in -- but if a clock on a proxy node is set an hour behind, one set an hour ahead and one set accurately, there is nothing we can do to prevent those nodes from dropping out or being out of sync.

the likely hood of clock skew causing node failure/dropout is likely a smaller issue than the resource contention issue we discussed on the zoom call monday.  there are currently a few different places where the node list can be inconsistent between proxy instances.

in lieu of doing an actual clustering protocol these corner cases will continue to exist.

Reply:
In the design doc, the algorithm we describe here is not depends on system clock.

We should have another meeting to make that part clear

Reply:
Relying on _any_ system clock is inherently a faulty idea (unless we are relying on a _single_ system clock, which given our current parameters is impossible).

We have changed how this algorithm works, I suggest reading the `syncing.md` doc included in here @zfy0701